### PR TITLE
ESP32: Disable random MAC address by default

### PR DESCRIPTION
**Pull request by sweetymhaiske**
#### Problem
* CHIP always use random MAC address for an AP interface

#### Change overview
* Add config option which when enabled then the AP interface will have a device MAC address as some customer protocol with SoftAP, needs to use Universally Administered MAC address.

#### Testing
* Build, flash, and monitored MAC address for all-clusters-app with Rendezvous mode as SoftAP

### CODE REVIEW

File: src/platform/ESP32/PlatformManagerImpl.cpp [hunk 1]
```diff
@@ -97,6 +99,7 @@ CHIP_ERROR PlatformManagerImpl::_InitChipStack(void)
     esp_wifi_get_mode(&mode);
```
Comment (dhrishi) [inline src/platform/ESP32/PlatformManagerImpl.cpp:99]: Shouldn't this be under the ifdef as well? Also, `wifi_mode_t mode;` above

Comment (sweetymhaiske) [reply src/platform/ESP32/PlatformManagerImpl.cpp:100]: Done

File: config/esp32/components/chip/Kconfig [hunk 2]
```diff
@@ -494,6 +494,12 @@ menu "CHIP Device Layer"
                 The amount of time (in milliseconds) after which the CHIP platform will deactivate the soft-AP
                 if it has been idle.
 
+        config WIFI_AP_RANDOM_MAC
+            bool "Enable Random MAC address"
+            default n
```
Comment (bzbarsky-apple) [inline config/esp32/components/chip/Kconfig:499]: Shouldn't the default be "yes"?  The spec-required behavior is the "yes" behavior....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
That said, it's not clear to me why we are adding this spec-violating configuration.   The PR description should explain why, at the very least....

File: config/esp32/components/chip/Kconfig (hunk 2 above)
Comment (sweetymhaiske) [reply config/esp32/components/chip/Kconfig]: Done

**Comment by sweetymhaiske** [comment]
> That said, it's not clear to me why we are adding this spec-violating configuration. The PR description should explain why, at the very least....

Updated.

File: src/platform/ESP32/PlatformManagerImpl.cpp [hunk 3]
```diff
@@ -94,6 +92,9 @@ CHIP_ERROR PlatformManagerImpl::_InitChipStack(void)
         goto exit;
     }
 
+#if defined(CONFIG_RENDEZVOUS_MODE_SOFTAP) && defined(CONFIG_WIFI_AP_RANDOM_MAC)
```
Comment (dhrishi) [inline src/platform/ESP32/PlatformManagerImpl.cpp:95]: @sweetymhaiske I think `CONFIG_RENDEZVOUS_MODE_SOFTAP` is from the examples. Ideally, we should use the chosen commissioning mode (flag/variable) that is available in the sdk

**Comment by dhrishi** [comment]
>That said, it's not clear to me why we are adding this spec-violating configuration. The PR description should explain why, at the very least....

@sweetymhaiske @bzbarsky-apple If it is mandatory in the spec, then I am not sure if we should do this. Although, we can add a check to do it only when commissioning mode is SoftAP (currently this happens for BLE as well). In this way, the usecase of a customer using their own protocol on SoftAP side while using commissioning over BLE works fine.

**Comment by bzbarsky-apple** [comment]
>  Although, we can add a check to do it only when commissioning mode is SoftAP

Just to make sure we are talking about the same things...

What the spec requires in "5.4.2.6.2. AP Operating Parameters" is that the BSSID:

> SHALL be randomly generated on each boot for privacy/tracking reasons. Broadcast bit SHALL be clear, Locally-administered bit SHALL BE set.

My understanding was that the AP MAC and the BSSID are (typically?) the same thing, but maybe I am misunderstanding the situation?   The spec does not seem to say anything specific about the actual MAC of the AP, just the BSSID.  Which of those does the PR affect?  Does ESP32 allow setting the two independently?

> (currently this happens for BLE as well).

Again, to make sure I understand this: does this mean we are currently using non-random identifiers for BLE, or is this a comment about the fact that this PR used to do a non-randomized MAC in station mode even if the device was configured for BLE, not Soft-AP, discovery?

I'm not sure anything in the spec talks about the MAC that needs to be used in station mode, offhand.

**Comment by dhrishi** [comment]
>My understanding was that the AP MAC and the BSSID are (typically?) the same thing

Yes, AFAIK that's the same. And ESP-IDF provides a single API which is to set the MAC address. Note that each interface (SoftAP, STA, Bluetooth) has a MAC address of its own .

So, as per the spec, ideally we should set random BSSID (and hence MAC) for the SoftAP interface when we use commissioning over SoftAP (device advertises its commissionable state using Wi-Fi SoftAP).

>Again, to make sure I understand this: does this mean we are currently using non-random identifiers for BLE, or is this a comment about the fact that this PR used to do a non-randomized MAC in station mode even if the device was configured for BLE, not Soft-AP, discovery?

Before this PR, even though commissioning over BLE (and not SoftAP) is used, SoftAP interface BSSID (MAC) was set randomly on every boot. This isn't required.

**Comment by bzbarsky-apple** [comment]
> Before this PR, even though commissioning over BLE (and not SoftAP) was used, SoftAP interface BSSID (MAC) was set randomly on every boot. This isn't required.

It's not required in the sense that the spec can't define what the SoftAP is doing if the SoftAP is not tryingto implement the spec, but it does undermine the spec's attempted privacy protections...  On the other hand, if the SoftAP is not trying to implement the spec, presumably it also doesn't follow the spec's time limits on beaconing, the spec's SSID format, etc?

**Comment by dhrishi** [comment]
>It's not required in the sense that the spec can't define what the SoftAP is doing if the SoftAP is not trying to implement the spec, but it does undermine the spec's attempted privacy protections

I agree to this in a way. But, there can be cases wherein I am using the SoftAP interface with my own protocol (which may have some other requirements) while continuing to use Commissioning over BLE. So, I think, we should not enforce a specific behavior.

> On the other hand, if the SoftAP is not trying to implement the spec, presumably it also doesn't follow the spec's time limits on beaconing, the spec's SSID format, etc?

I think so, if above (or similar) is the case and provided the other SoftAP parameter values mentioned in the spec are not generically used for different Wi-Fi usecases.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/platform/ESP32/CHIPDevicePlatformConfig.h [hunk 4]
```diff
@@ -78,3 +78,5 @@
 #define CHIP_DEVICE_CONFIG_EVENT_LOGGING_INFO_BUFFER_SIZE CONFIG_EVENT_LOGGING_INFO_BUFFER_SIZE
 #define CHIP_DEVICE_CONFIG_EVENT_LOGGING_DEBUG_BUFFER_SIZE CONFIG_EVENT_LOGGING_DEBUG_BUFFER_SIZE
 #define CHIP_DEVICE_CONFIG_LOG_PROVISIONING_HASH CONFIG_LOG_PROVISIONING_HASH
+#define CHIP_DEVICE_CONFIG_RANDOM_MAC CONFIG_WIFI_AP_RANDOM_MAC
+#define CHIP_DEVICE_RENDEZVOUS_MODE_SOFTAP (CONFIG_RENDEZVOUS_MODE == 1)
```
Comment (dhrishi) [inline src/platform/ESP32/CHIPDevicePlatformConfig.h:82]: @sweetymhaiske With this change, how will it work for `examples/temperature-measurement-app/esp32/` that does not have `CONFIG_RENDEZVOUS_MODE`?

**Pull request closed without merging**